app/src/main/java/com/boardgamegeek/ui/CollectionFragment.java

```diff
@@ -654,18 +658,74 @@ public void addFilter(@NonNull CollectionFilterer filter) {
 	@DebugLog
 	private void setEmptyText() {
 		if (emptyButton == null) return;
-		@StringRes int resId = R.string.empty_collection;
-		if (SyncPrefs.getLastCompleteCollectionTimestamp(getContext()) == 0L) {
-			resId = R.string.empty_collection_sync_never;
-			emptyButton.setVisibility(View.GONE);
-		} else if (hasFiltersApplied()) {
-			resId = R.string.empty_collection_filter_on;
-			emptyButton.setVisibility(View.VISIBLE);
-		} else if (!PreferencesUtils.isCollectionSetToSync(getContext())) {
-			resId = R.string.empty_collection_sync_off;
-			emptyButton.setVisibility(View.VISIBLE);
+		final Set<String> syncedStatuses = PreferencesUtils.getSyncStatuses(getContext());
+		final boolean collectionSyncOn = syncedStatuses != null && syncedStatuses.size() != 0;
+		if (collectionSyncOn) {
+			final boolean noPreviousSync = SyncPrefs.getLastCompleteCollectionTimestamp(getContext()) == 0L;
+			final boolean filterApplied = hasFiltersApplied();
+			if (noPreviousSync) {
+				setEmptyStateForNoAction(R.string.empty_collection_sync_never);
+			} else if (filterApplied) {
+				if (isAtLeastOneSyncOff(syncedStatuses, getListOfVisibleStatuses())) {
+					setEmptyStateForSettingsAction(R.string.empty_collection_filter_on_sync_partial);
+				} else {
+					setEmptyStateForReSyncAction(R.string.empty_collection_filter_on);
+				}
+			} else {
+				setEmptyStateForReSyncAction(R.string.empty_collection);
+			}
+		} else {
+			setEmptyStateForSettingsAction(R.string.empty_collection_sync_off);
+		}
+	}
+
+	private Set<String> getListOfVisibleStatuses() {
+		final Set<String> visibleStatuses = new HashSet<>(10);
+		for (CollectionFilterer filter : filters) {
+			if (filter instanceof CollectionStatusFilterer) {
+				CollectionStatusFilterer statusFilter = (CollectionStatusFilterer) filter;
+				visibleStatuses.addAll(statusFilter.getSelectedStatusesSet());
+			}
+		}
+		return visibleStatuses;
+	}
+
+	private boolean isAtLeastOneSyncOff(Set<String> syncedStatuses, Set<String> statusesToCheck) {
+		for (String status : statusesToCheck) {
+			if (!syncedStatuses.contains(status)) {
+				return true;
+			}
 		}
-		setEmptyText(getString(resId));
+		return false;
+	}
+
+	private void setEmptyStateForSettingsAction(@StringRes int textResId) {
+		setEmptyText(getString(textResId));
+		emptyButton.setOnClickListener(new OnClickListener() {
+			@Override
+			public void onClick(View v) {
+				startActivity(SettingsActivity.Companion.newIntent(getContext()));
```

Comment:
I've found this pattern to work very well when you add intent extras, since you can pass them as method params without the need to expose constants from the called \`Activity`. For example:
`SettingsActivity.Companion.newIntent(getContext(), SYNC)`
To point directly to the sync settings screen, that is what I meant to do in the next PR since at the moment it just starts the generic settings screen. Thoughts?

Reply:
I see. Where does the `SYNC` parameter come from? Is that a constant on `SettingsActivity`? If we keep this, I would like us to use `context.intentFor<SettingsActivity>` in SettingsActivity and mark the method as `@JvmStatic` to remove the `Companion`.

Reply:
Yeah, `SYNC` would be a constant defined within the `SettingsActivity`. The `@JvmStatic` annotation is a good idea, I don't like the `Companion` object either. How would the definition of `context.intentFor<SettingsActivity>` look?

Reply:
I'm okay with that your approach, as long as you get rid of the Companion reference here.